test/lib/jdk/test/lib/util/ForceGC.java

```diff
@@ -23,68 +23,57 @@
-import java.util.concurrent.TimeUnit;
+import java.lang.ref.PhantomReference;
+import java.lang.ref.Reference;
+import java.lang.ref.ReferenceQueue;
 import java.util.function.BooleanSupplier;
 
 /**
  * Utility class to invoke System.gc()
  */
 public class ForceGC {
-    private final static Cleaner cleaner = Cleaner.create();
-
-    private final CountDownLatch cleanerInvoked;
-    private Object o;
-    private int gcCount = 0;
-
-    public ForceGC() {
-        this.o = new Object();
-        this.cleanerInvoked = new CountDownLatch(1);
-        cleaner.register(o, cleanerInvoked::countDown);
-    }
-
-    private void doit(int iter) {
-        try {
-            for (int i = 0; i < 10; i++) {
-                System.gc();
-                gcCount++;
-                if (cleanerInvoked.await(100L, TimeUnit.MILLISECONDS)) {
-                    return;
-                }
-            }
-        } catch (InterruptedException unexpected) {
-            throw new AssertionError("unexpected InterruptedException");
-        }
-    }
+    // The jtreg testing timeout factor.
+    private static final double TIMEOUT_FACTOR = Double.valueOf(
+            System.getProperty("test.timeout.factor", "1.0"));
 
     /**
-     * Causes the current thread to wait until the {@code BooleanSupplier} returns true,
-     * unless the thread is interrupted or a predefined waiting time elapses.
+     * Causes the current thread to wait until the {@code booleanSupplier}
+     * returns true, or a specific waiting time elapses.  The waiting time
+     * is 1 second scalled with the jtreg testing timeout factor.
      *
-     * @param s boolean supplier
-     * @return true if the {@code BooleanSupplier} returns true and false if
-     *         the predefined waiting time elapsed before the count reaches zero.
-     * @throws InterruptedException if the current thread is interrupted while waiting
+     * @param booleanSupplier boolean supplier
+     * @return true if the {@code booleanSupplier} returns true, or false
+     *     if did not complete after the specific waiting time.
      */
-    public boolean await(BooleanSupplier s) {
-        o = null; // Keep reference to Object until now, to ensure the Cleaner
-                  // doesn't count down the latch before await() is called.
-        for (int i = 0; i < 10; i++) {
-            if (s.getAsBoolean()) {
-                System.out.println("ForceGC condition met after System.gc() " + gcCount + " times");
+    public static boolean wait(BooleanSupplier booleanSupplier) {
+        long timeout = Math.round(1000L * TIMEOUT_FACTOR);
+
+        ReferenceQueue<Object> queue = new ReferenceQueue<>();
+        Object obj = new Object();
+        PhantomReference<Object> ref = new PhantomReference<>(obj, queue);
+        obj = null;
+        Reference.reachabilityFence(obj);
+        Reference.reachabilityFence(ref);
+
+        for (int retries = (int)(timeout / 200); retries >= 0; retries--) {
```

Comment:
The logic around the timeout might be clearer if it was only based on the number of retries,
and can be scaled by the TIMEOUT_FACTOR too.

Reply:
I just curious if the factor could be set to some unusual values like "1.25".  Scaling on timeout, and then calculating the retires could be more accuracy for such circumstances, although it may be not necessary.  I moved the retries calculation close to the for-loop.  Hope it is better for readers.

Reply:
Yes, `Utils.adjustTimeout(long x)` uses floating point and then converts back.